Userland/Libraries/LibWeb/Layout/FormattingContext.cpp

```diff
@@ -1640,11 +1640,24 @@ bool FormattingContext::should_treat_height_as_auto(Box const& box, AvailableSpa
     if (computed_height.is_min_content() || computed_height.is_max_content() || computed_height.is_fit_content())
         return true;
 
+    // https://www.w3.org/TR/CSS2/visudet.html#propdef-height
+    // The percentage is calculated with respect to the height of the generated box's containing block.
     if (box.computed_values().height().contains_percentage()) {
-        if (available_space.height.is_max_content())
-            return true;
-        if (available_space.height.is_indefinite())
+
+        // If the height of the containing block is not specified explicitly (i.e., it depends on content height),
+        // and this element is not absolutely positioned, the value computes to 'auto'.
+        if (!box.is_absolutely_positioned() && should_treat_height_as_auto(*box.non_anonymous_containing_block()))
+        {
+            // NOTE: In quirks mode, the html element's height matches the viewport so it can be treated as definite
+            auto box_is_root_an_html_element_in_quirks_mode = box.document().in_quirks_mode()
+                && box.non_anonymous_containing_block()->dom_node()
+                && box.non_anonymous_containing_block()->dom_node()->is_html_html_element();
+
+            if (box_is_root_an_html_element_in_quirks_mode)
+                return false;
```

Comment:
This feels very weirdly factored. I would think that we mark the `html` element's corresponding layout node as having definite height at some earlier point. We shouldn't have to look at the document's quirks mode state here.

Does `m_state.get(*box.non_anonymous_containing_block()).has_definite_height()` not return true here?

Reply:
marking html element's height as definite should also allow to simplify code here https://github.com/SerenityOS/serenity/blob/master/Userland/Libraries/LibWeb/Layout/BlockFormattingContext.cpp#L648-L656